### PR TITLE
[NOTIFICATIONS #3] refactor(orders): unify orders notifications handlers

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

In my [first PR](https://github.com/cowprotocol/cowswap/pull/3902) where I've added a listener for `ON_POSTED_ORDER ` event and displayed `PendingOrderNotification`, everything was smooth and easy.
 `PendingOrderNotification` is a pure component and everything it needs to display a notification can be sent along with `ON_POSTED_ORDER` event just from the place where it was emitted.
It also allowed us to create a toast content using pure function `getPendingOrderNotificationToast()`.

But... it doesn't work like that with `FulfilledOrderNotification`.
In short, to display the notification we need:
1. Order info from redux store (`useOrder()`)
2. Surplus info (`useGetSurplusData()`)

Having this, `FulfilledOrderNotification` can't be pure and we can't derive all necessary information from the place where `ON_FULFILLED_ORDER` was emitted.

This problem made we to do following things:
1.  Make `FulfilledOrderNotification` as a container/smart
2. Add `getToastMessageCallback()` to derive plain text for `ON_TOAST_MESSAGE` event. Basically, I've added a callback as a `ref` to the notification component and use its `textContent` as a payload for `ON_TOAST_MESSAGE` event

Maybe this is not the best solution, but this is the only I have right now.

## Refactoring

1. Renamed `SWAP_POSTED_API` to `ORDER_CREATED` because we need to emit the event not only for swaps, but for other trades. To be able to distinguish orders I've added `UiOrderType` property to the event payload
2. Renamed `SWAP_TRADE_EXECUTED` to `ORDER_FULFILLED`, I think it reflects the sense of the event more clearly
3. Extracted `ORDERS_NOTIFICATION_HANDLERS` as a configuration object from `OrdersNotificationsUpdater` to make it more type safe and extendable
4. Removed `getPendingOrderNotificationProps()` and moved its logic into `PendingOrderNotification`

One special point I want to highlight is `OnFulfilledOrderPayload`.
[I've changed](https://github.com/cowprotocol/cowswap/pull/3921/files#diff-58c8547edd16d894df3504884dcfb2c3a04f0e4c8280d412f64fddce04fdcb19R35) its content by adding `order: EnrichedOrder`. I think, in generic services like the events library we should operate by generic entities (`EnrichedOrder` comes from SDK with is a core tool). In this context we should perceive `EnrichedOrder` as a `DTO`.

  # To Test

See https://github.com/cowprotocol/cowswap/pull/3934

### CODE REVIEW

File: apps/cowswap-frontend/src/modules/orders/pure/PendingOrderNotification/index.tsx [hunk 1]
```diff
@@ -105,24 +63,33 @@ export function PendingOrderNotification(props: PendingOrderNotificationProps) {
     },
   }
 
+  const ref = useCallback(
+    (node: HTMLDivElement) => {
+      if (node) onToastMessage(node.innerText)
+    },
+    [onToastMessage]
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/modules/orders/pure/PendingOrderNotification/index.tsx:70]: Please forgive my ignorance. How does this work?
When will this trigger? When the component is rendered?

File: apps/cowswap-frontend/src/common/hooks/useGetSurplusFiatValue.ts [hunk 2]
```diff
@@ -13,15 +13,15 @@ import { useSafeMemo } from 'common/hooks/useSafeMemo'
 import { getExecutedSummaryData } from 'utils/getExecutedSummaryData'
 import { ParsedOrder } from 'utils/orderUtils/parseOrder'
 
-type Output = {
+type SurplusData = {
```
Comment (anxolin) [inline apps/cowswap-frontend/src/common/hooks/useGetSurplusFiatValue.ts:16]: way better name I agree :)

File: apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx [hunk 3]
```diff
@@ -1,22 +1,45 @@
...
 interface FulfilledOrderNotificationProps {
   uid: string
   chainId: SupportedChainId
+  onToastMessage(message: string): void
 }
-export function FulfilledOrderNotification({ chainId, uid }: FulfilledOrderNotificationProps) {
+export function FulfilledOrderNotification({ chainId, uid, onToastMessage }: FulfilledOrderNotificationProps) {
+  const ref = useCallback(
+    (node: HTMLDivElement) => {
+      if (node) onToastMessage(node.innerText)
```
[the remark this answers is not in the document]
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx:22]: Ohh, this is what I wanted to do the other day, to try to not duplicate the content. 
I forgot that `innerText` was doing this.

@alfetopito, u remember I wanted to also get the text of the toast?

image

It feels a bit hacky, but I also don't like duplication of code, so I'm fine :P

File: apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx [hunk 4]
```diff
@@ -1,22 +1,45 @@
...
 
-  return <ExecutedSummary order={parsedOrder} />
+  const tx = {
+    hash: uid,
+    hashType: HashType.ETHEREUM_TX,
+  }
+
+  return (
+    <>
+      <ExecutedSummary ref={ref} order={parsedOrder} />
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx:39]: could you give a better name to the ref? `executedSummaryRef`?

File: apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx [hunk 5]
```diff
@@ -1,22 +1,45 @@
...
 import { ExecutedSummary } from '../ExecutedSummary'
 
 interface FulfilledOrderNotificationProps {
   uid: string
   chainId: SupportedChainId
+  onToastMessage(message: string): void
 }
-export function FulfilledOrderNotification({ chainId, uid }: FulfilledOrderNotificationProps) {
+export function FulfilledOrderNotification({ chainId, uid, onToastMessage }: FulfilledOrderNotificationProps) {
+  const ref = useCallback(
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx:20]: Is this a ref or a callback? how do you make sure this is only triggered once and when the moment is right? Feels a bit hacky

File: apps/cowswap-frontend/src/modules/orders/pure/PendingOrderNotification/index.tsx (hunk 1 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/modules/orders/pure/PendingOrderNotification/index.tsx:70]: I'm also confused. Also, there's a bit of repetition. Once we agree if this is a good pattern, if we keep it, I would suggest to build a custom hook

**Review by anxolin — COMMENTED** [review]
Soft approve

File: apps/cowswap-frontend/src/modules/orders/pure/PendingOrderNotification/index.tsx (hunk 1 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/orders/pure/PendingOrderNotification/index.tsx:70]: @anxolin I've got rid of the code duplications in the last PR: https://github.com/cowprotocol/cowswap/pull/3934

Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/orders/pure/PendingOrderNotification/index.tsx:70]: >How does this work?

>React will call the ref callback with the DOM element when the component mounts, and call it with null when it unmounts

https://legacy.reactjs.org/docs/refs-and-the-dom.html#callback-refs

File: apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx (hunk 4 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx:39]: It's not a case anymore, deleted this code in the last PR

File: apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx (hunk 5 above)
Comment (shoom3301) [reply apps/cowswap-frontend/src/modules/orders/containers/FulfilledOrderNotification/index.tsx:20]: https://legacy.reactjs.org/docs/refs-and-the-dom.html#callback-refs

>React will call the ref callback with the DOM element when the component mounts, and call it with null when it unmounts. Refs are guaranteed to be up-to-date before componentDidMount or componentDidUpdate fires.

**Pull request merged**